src/rt/rwlock.rs

```diff
@@ -182,22 +182,28 @@ impl RwLock {
             let thread_id = execution.threads.active_id();
 
             // Set the lock to the current thread
-            state.lock = match &state.lock {
+            let mut already_locked = false;
+            state.lock = match state.lock.take() {
                 None => {
                     let mut threads: HashSet<thread::Id> = HashSet::new();
                     threads.insert(thread_id);
                     Some(Locked::Read(threads))
                 }
-                Some(Locked::Read(current)) => {
-                    // TODO: refactor this to not create a `new` HashSet
-                    let mut new: HashSet<thread::Id> = HashSet::new();
-                    new.extend(current);
-                    new.insert(thread_id);
-                    Some(Locked::Read(new))
+                Some(Locked::Read(mut threads)) => {
+                    threads.insert(thread_id);
+                    Some(Locked::Read(threads))
+                }
+                Some(Locked::Write(writer)) => {
+                    already_locked = true;
+                    Some(Locked::Write(writer))
                 }
-                Some(Locked::Write(_)) => return false,
```

Comment:
Why not just keep
```rust
                 Some(Locked::Write(_)) => return false,
```
and ignore
```rust

                 Some(Locked::Write(writer)) => {
                    already_locked = true;
```

Reply:
@pickfire because of the [call to `Option::take` on line 186](https://github.com/tokio-rs/loom/pull/88/commits/cd84f9dd31438c588f53ef507e50ac3b1f952efc#diff-e9106da480ca203b688aa13f79b66051R186) that would be incorrect. Since we've `take`n the previous value, `state.lock` is currently `None`. If we return false here, we leave the state as `None`, incorrectly releasing the lock.